onnx_models/utils.py

```diff
@@ -0,0 +1,185 @@
+    np.dtype("int32"): ("si32", "i"),
+    np.dtype("uint32"): ("ui32", "I"),
+    np.dtype("int16"): ("si16", "h"),
+    np.dtype("uint16"): ("ui16", "H"),
+    np.dtype("int8"): ("si8", "b"),
+    np.dtype("uint8"): ("ui8", "B"),
+    np.dtype("float64"): ("f64", "d"),
+    np.dtype("float32"): ("f32", "f"),
+    np.dtype("float16"): ("f16", "e"),
+    np.dtype("bool"): ("i1", "?"),
+}
+
+
+def convert_numpy_to_iree_type_string(ndarr: np.ndarray):
+    shape = "x".join(str(x) for x in ndarr.shape)
+    dtype = numpy_to_iree_dtype_map[ndarr.dtype]
+    if shape == "":
+        return dtype
+    return f"{shape}x{dtype}"
+
+
+def pack_ndarray_to_binary(ndarr: np.ndarray):
+    mylist = ndarr.flatten().tolist()
+    dtype = ndarr.dtype
+    bytearr = b""
+    if dtype in numpy_to_iree_dtype_map:
+        iree_dtype = numpy_to_iree_dtype_map[dtype][1]
+        bytearr = struct.pack(f"{len(mylist)}{iree_dtype}", *mylist)
+    else:
+        raise NotImplementedError(
+            f"Unsupported data type in pack_ndarray_to_binary(): '{dtype}'"
+        )
+    return bytearr
+
+
+###############################################################################
+# ONNX utilities
+###############################################################################
+
+
+def convert_proto_elem_type_to_iree_dtype(etype) -> str:
+    if etype == onnx.TensorProto.BOOL:
+        return "i1"
+    if etype == onnx.TensorProto.INT4 or etype == onnx.TensorProto.UINT4:
+        return "i4"
+    if etype == onnx.TensorProto.INT8 or etype == onnx.TensorProto.UINT8:
+        return "i8"
+    if etype == onnx.TensorProto.INT16 or etype == onnx.TensorProto.UINT16:
+        return "i16"
+    if etype == onnx.TensorProto.INT32 or etype == onnx.TensorProto.UINT32:
+        return "i32"
+    if etype == onnx.TensorProto.INT64 or etype == onnx.TensorProto.UINT64:
+        return "i64"
+    if etype == onnx.TensorProto.FLOAT16:
+        return "f16"
+    if etype == onnx.TensorProto.FLOAT:
+        return "f32"
+    if etype == onnx.TensorProto.DOUBLE:
+        return "f64"
+    if etype == onnx.TensorProto.COMPLEX64:
+        return "complex<f32>"
+    if etype == onnx.TensorProto.COMPLEX128:
+        return "complex<f64>"
+    if etype == onnx.TensorProto.BFLOAT16:
+        return "bf16"
+    if etype == onnx.TensorProto.FLOAT8E4M3FN:
+        return "f8e4m3fn"
+    if etype == onnx.TensorProto.FLOAT8E4M3FNUZ:
+        return "f8e4m3fnuz"
+    if etype == onnx.TensorProto.FLOAT8E5M2:
+        return "f8e5m2"
+    if etype == onnx.TensorProto.FLOAT8E5M2FNUZ:
+        return "f8e5m2fnuz"
+    raise NotImplementedError(
+        f"type conversion for '{etype}' enum value not implemented"
+    )
+
+
+def convert_node_arg_type_to_numpy_dtype(type: str):
+    # TODO(scotttodd): use onnx.TensorProto instead? enums > strings
+    if type == "tensor(float)":
+        return np.float32
+    raise NotImplementedError(f"type conversion for '{type}' not implemented")
```

Comment:
Seems like the type here is just a string, but I can't find a complete list of what types are available: https://onnxruntime.ai/docs/api/python/api_summary.html#onnxruntime.NodeArg 

I'm tempted to use the TensorProto enums, but that would then involve loading the model into memory again.